### PR TITLE
feat: Feature Flag based routing

### PR DESCRIPTION
**Pull request by saq7**
# Pull Request

## JIRA Ticket

[CZID-8731](https://czi-tech.atlassian.net/browse/CZID-8731)

## Description

*The intended change, motivation and consequences.*

*Please tag a teammate as primary reviewer.*

## Notes

*Optional observations, comments or explanations.*
*If implementing a new feature, note any relevant analytics added.*

## Tests

* *Server-side code should have automated tests*
* *Client-side code and scripts should have at least a manual test plan*

[CZID-8731]: https://czi-tech.atlassian.net/browse/CZID-8731?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Dockerfile
Comment (jfoo1984) [inline Dockerfile:11]: this is currently being run in the `entrypoint.sh`: https://github.com/chanzuckerberg/czid-graphql-federation-server/blob/9af53fe5d9b8c0f88659014e054c430f708e1cfd/entrypoint.sh#L2

[`npm run build` script defined in package.json](https://github.com/chanzuckerberg/czid-graphql-federation-server/blob/9af53fe5d9b8c0f88659014e054c430f708e1cfd/package.json#L7).  

Is there any reason we want to run build a separate time here?

I think we do want a single command to run build and start (which is why they're in a script together), to make sure the server is built with the latest schema.

File: docker-compose.yml
Comment (jfoo1984) [inline docker-compose.yml:13]: I don't know if we're looking to set this up in deployed environments at this time, but we'll eventually need to add vars in each github env in this, and configure our github actions to pick up the variables and set them with happy config https://github.com/chanzuckerberg/czid-graphql-federation-server/blob/9af53fe5d9b8c0f88659014e054c430f708e1cfd/.github/workflows/deploy-sandbox.yml#L34-L39

File: .meshrc.yaml
Comment (jfoo1984) [inline .meshrc.yaml:225]: I know this PR is WIP, but we do need to enable this before we can merge the PR.

**Comment by jfoo1984** [comment]
I believe the general approach that we decided on was that the feature flag will be passed as an argument from the client (React app), since it already retrieves the feature flags as part of loading pages.  Are there specific use cases we had in mind where graphQL mesh needs to fetch the feature flags from the CZ ID Rails app?

File: Dockerfile
Comment (saq7) [reply Dockerfile:11]: This can be removed I guess. I added it there because I was running into an issue where it complained that the assets weren't already built. This happened after I had deleted the build folder

File: docker-compose.yml
Comment (saq7) [reply docker-compose.yml:13]: Yeah, it makes sense to hold off deploying it but I added it here so that it is available for local development.

**Comment by saq7** [comment]
@jfoo1984 I don't think we have a specific usecase where the feature flag needs to be fetched by the federation server. Though I added it here to make it easier to develop and use as needed until the end-to-end flow wrt to the fetching of the feature flag is complete. We can remove this down the road.

File: utils/httpUtils.ts
Comment (jfoo1984) [inline utils/httpUtils.ts:107]: This function fetches feature flags from Rails right?  Do we need this?  I thought that we are expecting all feature flags to be passed from the front end.

File: .meshrc.yaml
Comment (jfoo1984) [inline .meshrc.yaml:22]: This is already configured above, as `CZIDENTITIES`.  i don't know if they are exactly the same, but we should figure out which one is most up to date.  I think @rzlim08 added the one above.

File: .meshrc.yaml
Comment (jfoo1984) [reply .meshrc.yaml:277]: bump on this, we need to uncomment this and make sure things work.

File: Dockerfile
Comment (jfoo1984) [reply Dockerfile:11]: bump, can this be removed without issue?

File: docker-compose.yml
Comment (jfoo1984) [reply docker-compose.yml:13]: Yea, we do have to have this for local dev regardless, but my comment pertained to deployed environments.  for dev and sandbox, #96 added these env vars as part of the TF config

File: utils/httpUtils.ts
Comment (jfoo1984) [inline utils/httpUtils.ts:78]: would it be viable to put this code in the `callGraphqlSource` util function instead, and call that util function here?  This util file is getting fairly long and complex, so it'd be nice to extract the chunks that make sense.

https://github.com/chanzuckerberg/czid-graphql-federation-server/blob/010967bd3011dd1544a6bad984f4e64e2f3f0b08/utils/callGraphqlSource.ts#L3

File: utils/httpUtils.ts
Comment (bchu1) [inline utils/httpUtils.ts:178]: Question - why are there 2 places where the `next_gen` flag could be?

File: utils/httpUtils.ts
Comment (bchu1) [inline utils/httpUtils.ts:28]: noticed that the other ones do `nextGenEnabled` first, might be nice to do that here too

File: utils/httpUtils.ts
Comment (jfoo1984) [inline utils/httpUtils.ts:68]: This is not the correct URL for nextGen, because this is concatenating the Rails base URL with the endpoint passed in, right?  Since we don't actually check the the flag until we get to this method, the resolvers won't actually know which URL to call.  Though for calling nextGen, since it's graphQL, I don't think we have to. worry about a different URL, it's always going to be `/graphql`.

**Comment by jfoo1984** [comment]
Can we add unit tests for all of the new code here as well?

File: utils/httpUtils.ts
Comment (jfoo1984) [reply utils/httpUtils.ts:178]: I thought we are planning to only get the feature flag via the headers, and not querying them from Rails.  Do we need that functionality at all?

Comment (jfoo1984) [reply utils/httpUtils.ts:178]: I just saw @saq7 's previous comment about this: https://github.com/chanzuckerberg/czid-graphql-federation-server/pull/44#issuecomment-1889595336

I'm skeptical we should add this unless we had a particular use in mind.  We can save off the code for fetching from Rails in a branch or wiki if we'd like, but I'd recommend we just stick to the header pattern and require clients to make requests that way, until we encounter a use case where that won't work.

**Review by rainandbare — COMMENTED** [review]
Have we tested to make sure that all the current rails calls are working as expected?

File: .meshrc.yaml
Comment (saq7) [reply .meshrc.yaml]: Thanks for the call out. I updated the details here in the two sources from Ryan's and removed that one. Will update the schemas as well

File: utils/httpUtils.ts
Comment (saq7) [reply utils/httpUtils.ts:178]: That is how the rails endpoint is set up. `launched_feature_list` is a universal list for all users and `allowed_feature_list` is user specific

File: utils/httpUtils.ts
Comment (saq7) [reply utils/httpUtils.ts]: sorry, oversight. Fixed.

File: utils/httpUtils.ts
Comment (saq7) [reply utils/httpUtils.ts:170]: we can rename it. But what other functionality are you envisioning?

File: utils/httpUtils.ts
Comment (saq7) [reply utils/httpUtils.ts]: I commented out that code for the merge. I suggest we leave the comments in place so that they can be enabled for local development until we are in the flow of building using this pattern

**Comment by saq7** [comment]
@rainandbare yes the current rails calls should work as expected. Let me know if you run into any issues

File: .meshrc.yaml
Comment (rzlim08) [inline .meshrc.yaml:20]: This should be the `workflows` schema

File: docker-compose.yml
Comment (rzlim08) [inline docker-compose.yml:13]: I don't think `NEXTGEN_API_URL` is necessary anymore

File: .meshrc.yaml
Comment (rzlim08) [inline .meshrc.yaml:6]: Could you delete this schema if it's not being used anymore?

File: utils/httpUtils.ts
Comment (jfoo1984) [reply utils/httpUtils.ts:170]: probably if we end up needing feature flags around writing to nextGen that are separate from the read feature flags

**Review by rainandbare — APPROVED** [review]

File: utils/httpUtils.ts
Comment (rainandbare) [reply utils/httpUtils.ts:170]: Let's merge this in for now and make those variable name changes with any tests that need to be written.

**Comment by jfoo1984** [comment]
To capture context here, I think we should test this PR integrated on sandbox and make sure it works before we merge it.  unfortunately, this was already merged, so it's now being reverted

File: utils/httpUtils.ts
Comment (jfoo1984) [reply utils/httpUtils.ts:170]: stated in a comment on the whole PR, but we don't need to rush to merge this.  we can (and should IMO) deploy this to sandbox and make sure the integration works, and clean things up before merging

**Pull request merged**
